### PR TITLE
feat: defer to prod changes

### PR DESCRIPTION
**Pull request by gaurpulkit**
#### Changes
- Updated all extension commands to show under `dbt Power User`
- Added new insights panel with defer, big query cost estimate and health checker modules
- Added new command `dbtPowerUser.buildCurrentProject` and an action button for the command with logs
- Used logging service for logging info, debug and error messages
- Added new defer button in bottom quick pick toolbar. This button will show defer status of current project
- Defer setting can be done using local or remote manifest using project integration via saas app
- Following commands in extension is updated to use defer config `run model`, `build model`, `build project`, `test model`

Screenshot:
image

#### How to test
- Learn about defer feature: https://infinitelambda.com/dbt-deferral-simplify-development/
- Open Insights panel or click defer button in bottom toolbar
- Select a model from dbt project (defer config will be saved to the project based on the active file selected)
- enable/disable defer switch
- select local or remote manifest path
- for remote manifest, make sure project integration is done via SaaS app
- toggle favor state as needed
- Run any of the following dbt command by using the action buttons: `run model`, `build model`, `build project`, `test model`
image
- Terminal should show the defer arguments with the command and models should be skipped based on the defer config settings

#### Usecases:
Different project setup in vscode workspace
Reference: https://github.com/microsoft/vscode/wiki/Adopting-Multi-Root-Workspace-APIs#settings
- [x] single dbt project opened as workspace
- [x] multiple dbt projects under same directory - open directory as workspace
- [x] multiple dbt projects from different directories - open workspace and add different dbt project as folder
   - [x] projects with same name from different directories under same workspace

## Checklist

- [x] I have run this code and it appears to resolve the stated issue
- [ ] `README.md` updated and added information about my change

### CODE REVIEW

File: src/webview_provider/altimateWebviewProvider.ts [hunk 1]
```diff
@@ -280,11 +286,16 @@ export class AltimateWebviewProvider implements WebviewViewProvider {
...
-              } https: data:; script-src 'nonce-${nonce}';">
+              <meta
+                http-equiv="Content-Security-Policy"
+                content="default-src 'none'; style-src 'unsafe-inline' ${
+                  webview.cspSource
+                }; img-src ${
+                  webview.cspSource
+                } https: data:; script-src 'nonce-${nonce}'; connect-src ${
+                  webview.cspSource
+                } https://data-pipeline-output-local-testing.s3.amazonaws.com;"
```
Comment (gaurpulkit) [inline src/webview_provider/altimateWebviewProvider.ts:297]: will revert it

File: src/altimate.ts [hunk 2]
```diff
@@ -369,4 +380,20 @@ export class AltimateRequest {
     });
     return (await response.json()) as Record<string, any> | undefined;
   }
+
+  async fetchProjectIntegrations() {
+    return this.fetch("dbt/v1/project_integrations");
```
Comment (mdesmet) [inline src/altimate.ts:385]: This should be typed.

File: src/altimate.ts [hunk 3]
```diff
@@ -369,4 +380,20 @@ export class AltimateRequest {
...
+  }
+
+  async downloadArtifect(
+    artifact_type: string,
+    dbt_core_integration_id: number,
+  ) {
+    return this.fetch(
```
Comment (mdesmet) [inline src/altimate.ts:392]: Also should be typed.

File: src/webview_provider/insightsPanel.ts [hunk 4]
```diff
@@ -118,6 +121,137 @@ export class InsightsPanel extends AltimateWebviewProvider {
     }
   }
 
+  private async fetchProjectIntegrations(syncRequestId: string | undefined) {
+    try {
+      console.log("Fetching project integrations");
+      const response = await this.altimateRequest.fetchProjectIntegrations();
+      if (!response) {
```
Comment (mdesmet) [inline src/webview_provider/insightsPanel.ts:128]: No response means bad credentials? There must be other error conditions as well.

File: webview_panels/src/modules/defer/DeferToProduction.tsx [hunk 5]
```diff
@@ -93,27 +111,80 @@ const DeferToProduction = (): JSX.Element => {
...
+      ...prevState,
+      manifestPathType: option,
+    }));
+  };
+
+  const handleIntegrationSelect = async (selectedOption: {
+    label: string;
+    value: number;
+  }) => {
+    const response = await executeRequestInSync("downloadManifest", {
```
Comment (mdesmet) [inline webview_panels/src/modules/defer/DeferToProduction.tsx:165]: The manifest should be downloaded prior to dbt command execution. We should guarantee that the latest artifact is used for any command execution.

File: src/webview_provider/insightsPanel.ts (hunk 4 above)
Comment (gaurpulkit) [reply src/webview_provider/insightsPanel.ts:128]: Yes, removing it as `fetch` function already handles the same thing.

File: webview_panels/src/modules/defer/DeferToProduction.tsx (hunk 5 above)
Comment (gaurpulkit) [reply webview_panels/src/modules/defer/DeferToProduction.tsx:165]: Yes. As soon as the user selects any of the project integrations, it will automatically download the latest artifact.

File: src/webview_provider/insightsPanel.ts [hunk 6]
```diff
@@ -118,6 +121,133 @@ export class InsightsPanel extends AltimateWebviewProvider {
...
+          body: {
+            response: [],
+          },
+          status: false,
+        },
+      });
+    }
+  }
+
+  async downloadManifestLocally(
```
Comment (AdiGajbhiye) [inline src/webview_provider/insightsPanel.ts:157]: move this function to `src/altimate.ts` for downloading files. So that it can be used in other places.

File: src/altimate.ts [hunk 7]
```diff
@@ -291,6 +326,48 @@ export class AltimateRequest {
...
+    url: string,
+    filePath = "manifest.json",
+  ): Promise<string> {
+    const tempFolder = os.tmpdir();
+
+    mkdir(tempFolder, { recursive: true }, (err) => {
+      console.log(err);
+    });
+
+    const destinationPath = join(tempFolder, filePath);
```
Comment (mdesmet) [inline src/altimate.ts:339]: This path might conflict if multiple projects are ran simultaneously. maybe better use a uuid?

File: src/altimate.ts [hunk 8]
```diff
@@ -291,6 +326,48 @@ export class AltimateRequest {
...
+      response.body?.pipe(fileStream);
+      response.body?.on("error", reject);
+      fileStream.on("finish", resolve);
+    });
+
+    console.log("File downloaded successfully!");
+    return destinationPath;
+  }
+
+  getQueryString = (params: Record<string, string | number>): string => {
```
Comment (mdesmet) [inline src/altimate.ts:360]: This method should be private.

File: src/dbt_client/dbtCoreIntegration.ts [hunk 9]
```diff
@@ -411,8 +420,56 @@ export class DBTCoreProjectIntegration
...
+      if (response?.url) {
+        const manifestPath = await this.altimateRequest.downloadFileLocally(
+          response.url,
+        );
+        console.log(`Set remote manifest path: ${manifestPath}`);
+        return [
+          "--defer",
+          "--state",
+          manifestPath,
+          favorState ? "--favor-state" : "",
```
Comment (mdesmet) [inline src/dbt_client/dbtCoreIntegration.ts:459]: Don't insert empty string.

File: src/dbt_client/dbtCoreIntegration.ts [hunk 10]
```diff
@@ -411,8 +420,56 @@ export class DBTCoreProjectIntegration
...
+        ];
+      }
+    }
+    return [
+      "--defer",
+      "--state",
+      manifestPathForDeferral,
+      favorState ? "--favor-state" : "",
```
Comment (mdesmet) [inline src/dbt_client/dbtCoreIntegration.ts:467]: Don't insert empty string

File: src/webview_provider/altimateWebviewProvider.ts [hunk 11]
```diff
@@ -185,6 +185,12 @@ export class AltimateWebviewProvider implements WebviewViewProvider {
             });
           }
           break;
+        case "showInfoNotification":
```
Comment (mdesmet) [inline src/webview_provider/altimateWebviewProvider.ts:188]: showInformationMessage. Let's not introduce new vocab.

File: webview_panels/src/modules/defer/DeferToProduction.tsx [hunk 12]
```diff
@@ -114,6 +131,62 @@ const DeferToProduction = (): JSX.Element => {
...
+      dbt_core_integration_id: selectedOption.value,
+    });
+    const data = response as {
+      url: string;
+      dbt_core_integration_file_id: number;
+    };
+    if (data.url === "" && data.dbt_core_integration_file_id === -1) {
+      executeRequestInAsync("showInfoNotification", {
+        infoMessage: `No remote manifest file present for dbt core integration: ${selectedOption.label}`,
+      });
```
Comment (mdesmet) [inline webview_panels/src/modules/defer/DeferToProduction.tsx:174]: Should we block on this? I think we should still updateDeferConfig.

**Review by mdesmet — COMMENTED** [review]
LGTM from code perspective. Just some small remarks.

File: src/altimate.ts (hunk 7 above)
Comment (gaurpulkit) [reply src/altimate.ts:339]: It automatically creates a uuid. It's not a fixed path

File: webview_panels/src/modules/defer/DeferToProduction.tsx (hunk 12 above)
Comment (gaurpulkit) [reply webview_panels/src/modules/defer/DeferToProduction.tsx:174]: There's no reason to update defer config if there's no project integration imo

File: webview_panels/src/modules/defer/DeferToProduction.tsx [hunk 13]
```diff
@@ -93,23 +125,102 @@ const DeferToProduction = (): JSX.Element => {
...
   };
 
-  const onManifestBlur = async () => {
-    const response = await executeRequestInSync("updateDeferConfig", {
-      key: "manifestPathForDeferral",
-      value: manifestPathForDeferral,
-      isPreviewFeature: true,
+  const onLocalManifestBlur = async () => {
+    if (!showManifestError) {
+      const response = await executeRequestInSync("updateDeferConfig", {
```
Comment (saravmajestic) [inline webview_panels/src/modules/defer/DeferToProduction.tsx:142]: you could modify `updateDeferConfig` to accept of of key value pairs to avoid multiplecalls

File: webview_panels/src/modules/defer/DeferToProduction.tsx [hunk 14]
```diff
@@ -93,23 +125,102 @@ const DeferToProduction = (): JSX.Element => {
...
+          }),
+        ),
+        manifestPathType: "remote",
+      }));
+      return;
+    }
+  };
+
+  const handleManifestPathTypeChange = async (option: string) => {
+    if (option === "remote") {
```
Comment (saravmajestic) [inline webview_panels/src/modules/defer/DeferToProduction.tsx:185]: better to an enum

File: webview_panels/src/modules/defer/DeferToProduction.tsx [hunk 15]
```diff
@@ -93,23 +125,102 @@ const DeferToProduction = (): JSX.Element => {
...
+
+  const handleIntegrationSelect = async (selectedOption: {
+    label: string;
+    value: number;
+  }) => {
+    const response = await executeRequestInSync("testRemoteManifest", {
+      dbt_core_integration_id: selectedOption.value,
     });
-    if (!(response as { updated: boolean }).updated) {
+    const data = response as {
```
Comment (saravmajestic) [inline webview_panels/src/modules/defer/DeferToProduction.tsx:201]: we should make `executeRequestInSync` to accept type generics to avoid typecasting everytime. I will take care of this separately

File: webview_panels/src/modules/defer/DeferToProduction.tsx [hunk 16]
```diff
@@ -93,23 +125,102 @@ const DeferToProduction = (): JSX.Element => {
...
+      url: string;
+      dbt_core_integration_file_id: number;
+    };
+    if (data.url === "" && data.dbt_core_integration_file_id === -1) {
+      executeRequestInAsync("showInformationMessage", {
+        infoMessage: `No remote manifest file present for dbt core integration: ${selectedOption.label}`,
+      });
+    } else {
```
Comment (saravmajestic) [inline webview_panels/src/modules/defer/DeferToProduction.tsx:209]: you could do early `return` for the `if` condition and avoid else block to make the code more maintainable and readable
```
if (condition){
return
}
do next steps
```

File: webview_panels/src/modules/defer/DeferToProduction.tsx [hunk 17]
```diff
@@ -140,25 +251,75 @@ const DeferToProduction = (): JSX.Element => {
...
               </Label>
             </FormGroup>
-            <FormGroup>
-              <Stack>
+            <Label>Save your file location</Label>
+            <FormGroup check className={classes.pathSelection}>
+              <div className={classes.pathSelectionRow}>
                 <Label
+                  check
                   for="manifestPath"
```
Comment (saravmajestic) [inline webview_panels/src/modules/defer/DeferToProduction.tsx:259]: this `for` attribute should match the id of the input radio element to trigger click event on label. Otherwise remove the for attribute and just leave the labelwrapping the input. This will give better user experience. currently when i click the label it does not select the radio button

File: webview_panels/src/modules/defer/DeferToProduction.tsx [hunk 18]
```diff
@@ -140,25 +251,75 @@ const DeferToProduction = (): JSX.Element => {
                 />
               </Label>
             </FormGroup>
-            <FormGroup>
-              <Stack>
+            <Label>Save your file location</Label>
```
Comment (saravmajestic) [inline webview_panels/src/modules/defer/DeferToProduction.tsx:254]: It would have been better to extract this whole manifest selection logic to a separate component to avoid bloating this defer component

File: webview_panels/src/modules/defer/defer.module.scss [hunk 19]
```diff
@@ -23,4 +23,26 @@
       }
     }
   }
+
+  .path-selection {
+    padding: 16px;
+    border-radius: 10px;
+    background: #1c1c1c;
```
Comment (saravmajestic) [inline webview_panels/src/modules/defer/defer.module.scss:30]: use it from css variables

**Review by saravmajestic — COMMENTED** [review]
- webview_panels/src/assets/icons/eye-yellow.svg - this file is not used anymore, you can remove this
- webview_panels/src/modules/dataPilot/sampleData.ts - this file can be deleted, not used anymore
- webview_panels/src/uiCore/UiToolkit.stories.tsx - can be deleted as it is repeated due to merge

File: src/dbt_client/dbtCoreIntegration.ts [hunk 20]
```diff
@@ -388,6 +406,58 @@ export class DBTCoreProjectIntegration
...
+    if (!manifestPathForDeferral) {
+      return [];
+    }
+    if (manifestPathType === "remote" && dbt_core_integration_id) {
+      const response = await this.altimateRequest.downloadArtifact(
+        "manifest",
+        dbt_core_integration_id,
+      );
+      if (response?.url) {
+        const manifestPath = await this.altimateRequest.downloadFileLocally(
```
[the remark this answers is not in the document]
Comment (gaurpulkit) [reply src/dbt_client/dbtCoreIntegration.ts:437]: Yeah, we discussed it. For now, we will be downloading the file for each execution so that it's always the latest one.

File: src/altimate.ts (hunk 7 above)
Comment (mdesmet) [reply src/altimate.ts:348]: I checked it, it is not a unique path. it is shared with other stuff.

File: src/commands/bigQueryCostEstimate.ts [hunk 21]
```diff
@@ -14,12 +14,14 @@ export class BigQueryCostEstimate {
     private telemetry: TelemetryService,
   ) {}
 
-  async estimateCost() {
+  async estimateCost({ returnResult }: { returnResult?: boolean }) {
```
Comment (mdesmet) [inline src/commands/bigQueryCostEstimate.ts:17]: How is this change related?

File: webview_panels/src/modules/dataPilot/datapilot.module.scss [hunk 22]
```diff
@@ -1,7 +1,7 @@
-body {
+body{
   padding: var(--spacing-3xl) !important;
 }
-.chat-list {
+.chat-list{
```
Comment (mdesmet) [inline webview_panels/src/modules/dataPilot/datapilot.module.scss:4]: Unrelated changes

**Review by mdesmet — COMMENTED** [review]
LGTM except some small remarks

File: src/commands/bigQueryCostEstimate.ts (hunk 21 above)
Comment (gaurpulkit) [reply src/commands/bigQueryCostEstimate.ts:17]: @saravmajestic

File: src/webview_provider/insightsPanel.ts [hunk 23]
```diff
@@ -0,0 +1,267 @@
...
+      if (!this.altimateRequest.handlePreviewFeatures()) {
+        throw new Error("Invalid credentials");
+      }
+
+      if (Object.keys(params).every((key) => /^\d+$/.test(key))) {
+        params = Object.values(params);
+      }
+
+      let updateConfigs: UpdateConfigProps[];
+      if (Array.isArray(params)) {
```
Comment (saravmajestic) [inline src/webview_provider/insightsPanel.ts:85]: instead of doing this, it will be better to restrict only to array types so we dont need to different cases

File: webview_panels/src/modules/app/requestExecutor.ts [hunk 24]
```diff
@@ -8,7 +8,7 @@ const requestMap: Record<
 
 export const executeRequestInSync = (
   url: string,
-  params: Record<string, unknown>,
+  params: Record<string, unknown> | Record<string, unknown>[],
```
Comment (saravmajestic) [inline webview_panels/src/modules/app/requestExecutor.ts:11]: if this is for updating defer config, pls keep as Record<string, unknown> and pass the values as {config: [{key: value}]}

File: webview_panels/src/modules/defer/DeferToProduction.tsx [hunk 25]
```diff
@@ -10,29 +9,237 @@ import {
...
+interface ManifestSelectionProps {
+  manifestPathForDeferral: string;
+  manifestPathType: ManifestPathType;
+  projectIntegrations: DropdownOptions[];
+  dbt_core_integration_id: number;
+  setDeferState: React.Dispatch<React.SetStateAction<DeferToProductionProps>>;
+  setProjectIntegrations: () => Promise<void>;
 }
+
+const ManifestSelection = ({
```
Comment (saravmajestic) [inline webview_panels/src/modules/defer/DeferToProduction.tsx:46]: as a better practice, pls keep each component in separate file

**Review by saravmajestic — COMMENTED** [review]
minor commens

**Comment by gaurpulkit** [comment]
Might need to change the function getProjectRelativePath. @saravmajestic @mdesmet

**Comment by gaurpulkit** [comment]
@saravmajestic, @mdesmet wants to change the input of the local manifest file to a file selector.

File: src/commands/bigQueryCostEstimate.ts (hunk 21 above)
Comment (saravmajestic) [reply src/commands/bigQueryCostEstimate.ts:17]: @mdesmet this PR was supposed to have bigquery cost estimate module as well in the insights panel. That was missing after rebasing with core integration refactor. I have added it again as it is part of insights panel

**Comment by saravmajestic** [comment]
> @saravmajestic, @mdesmet wants to change the input of the local manifest file to a file selector.

Implemented folder selector for the manifest path selection

File: package.json [hunk 26]
```diff
@@ -248,16 +280,30 @@
...
+          "id": "dbtPowerUser.Insights",
+          "name": "",
+          "type": "webview"
+        }
       ]
     },
     "commands": [
       {
         "command": "dbtPowerUser.puQuickPick",
+        "category": "dbt Power User",
```
Comment (saravmajestic) [inline package.json:295]: added `category` for all the commands to show under `dbt Power User` label in `cmd+shift+P` menu

File: src/commands/bigQueryCostEstimate.ts (hunk 21 above)
Comment (mdesmet) [inline src/commands/bigQueryCostEstimate.ts:17]: Why do we need to have returnResult, can't we not simply ignore it if we don't need it?

File: src/dbt_client/dbtCoreIntegration.ts [hunk 27]
```diff
@@ -398,6 +426,123 @@ export class DBTCoreProjectIntegration
...
+    }
+
+    const validation = await this.altimateRequest.validateCredentials(
+      config.instance,
+      config.key,
+    );
+
+    if (!validation?.ok) {
+      window.showErrorMessage(
+        "Cannot run with defer to prod set up as Auth tokens are invalid",
```
Comment (mdesmet) [inline src/dbt_client/dbtCoreIntegration.ts:447]: Has this error message been reviewed by @ppradnesh and @anandgupta42 ?

Comment (mdesmet) [reply src/dbt_client/dbtCoreIntegration.ts:447]: Also won't this result in mutliple error messages as we throw the error hereunder?

File: src/dbt_client/dbtCoreIntegration.ts [hunk 28]
```diff
@@ -50,6 +54,14 @@ interface ResolveReferenceSourceResult {
   identifier: string;
 }
 
+interface DeferConfig {
+  deferToProduction: boolean;
+  favorState: boolean;
+  manifestPathForDeferral: string;
+  manifestPathType?: ManifestPathType;
+  dbt_core_integration_id?: number;
```
Comment (mdesmet) [inline src/dbt_client/dbtCoreIntegration.ts:62]: Usually we use camelcase for internal config. snake case is for python and api interfaces.

File: src/statusbar/deferToProductionStatusBar.ts [hunk 29]
```diff
@@ -0,0 +1,116 @@
...
+    const currentProjectRoot = await this.getCurrentProjectRoot();
+    const currentConfig: Record<string, DeferConfig> = await workspace
+      .getConfiguration("dbt", currentDocument?.uri)
+      .get("deferConfigPerProject", {});
+
+    if (currentConfig[currentProjectRoot]?.deferToProduction) {
+      this.showTextInStatusBar("$(sync) Defer");
+      return;
+    }
+    this.showTextInStatusBar("$(sync-ignored) Defer");
```
Comment (mdesmet) [inline src/statusbar/deferToProductionStatusBar.ts:114]: I think this shouldn't show if no project can be determined.

**Review by mdesmet — COMMENTED** [review]
overall LGTM except some minor remarks

File: src/commands/bigQueryCostEstimate.ts (hunk 21 above)
Comment (saravmajestic) [reply src/commands/bigQueryCostEstimate.ts:17]: @mdesmet This change was needed in bigquery cost estimator in insights panel to show the response.

File: src/dbt_client/dbtCoreIntegration.ts (hunk 27 above)
Comment (saravmajestic) [reply src/dbt_client/dbtCoreIntegration.ts:447]: @suryaiyer95

Comment (mdesmet) [reply src/dbt_client/dbtCoreIntegration.ts:485]: But won't it result in an unhandled exception?

File: src/dbt_client/dbtCoreIntegration.ts (hunk 28 above)
Comment (mdesmet) [reply src/dbt_client/dbtCoreIntegration.ts:62]: For clarity here it used as an internal config, not an API or python method.

File: src/commands/bigQueryCostEstimate.ts (hunk 21 above)
Comment (mdesmet) [reply src/commands/bigQueryCostEstimate.ts:17]: I repeat my question, why not always return the result and get rid of the parameter?

File: src/statusbar/deferToProductionStatusBar.ts (hunk 29 above)
Comment (saravmajestic) [reply src/statusbar/deferToProductionStatusBar.ts]: made changes

File: src/dbt_client/dbtCoreIntegration.ts (hunk 28 above)
Comment (saravmajestic) [reply src/dbt_client/dbtCoreIntegration.ts]: fixed

File: src/webview_provider/insightsPanel.ts [hunk 30]
```diff
@@ -0,0 +1,465 @@
...
+    try {
+      this.dbtTerminal.debug("InsightsPanel", "Updating defer config", params);
+      if (!params.projectRoot) {
+        window.showErrorMessage("Please select a project");
+        return;
+      }
+      // defer config is preview feature, then check keys
+      if (!this.altimateRequest.handlePreviewFeatures()) {
+        throw new Error("Invalid credentials");
+      }
```
Comment (mdesmet) [inline src/webview_provider/insightsPanel.ts:122]: It should be allowed to choose the local manifest file and deactivate the defer to prod feature in case the user doesn't have an altimate key anymore (previously having setup it).

File: src/webview_provider/insightsPanel.ts [hunk 31]
```diff
@@ -0,0 +1,465 @@
...
+      const response = await this.altimateRequest.fetchArtifactUrl(
+        "manifest",
+        dbtCoreIntegrationId,
+      );
+      // TODO: Can be removed once the backend sends a 404
+      if (response.url === "" && response.dbt_core_integration_file_id === -1) {
+        throw new Error(
+          "No remote manifest file present for selected dbt core integration",
+        );
+      }
```
Comment (mdesmet) [inline src/webview_provider/insightsPanel.ts:263]: This should be handled through the exception thrown in altimate fetch method. Needs to be tested and better error message be displayed
